dfc.providerportal.findanapprenticeship/src/dfc.providerportal.findanapprenticeship/Functions/GenerateProviderExportFunction.cs

```diff
@@ -28,53 +34,158 @@ public GenerateProviderExportFunction(IApprenticeshipService apprenticeshipServi
         }
 
         [FunctionName("GenerateProviderExport")]
-        public async Task Run([TimerTrigger("%GenerateProviderExportSchedule%")]TimerInfo timer, ILogger log, CancellationToken ct)
+        public Task Run([TimerTrigger("%GenerateProviderExportSchedule%")]TimerInfo timer, ILogger log, CancellationToken ct)
         {
-            var exportKey = ExportKey.FromUtcNow();
+            return GenerateProviderExport(log, ct);
+        }
+
+        [FunctionName("GenerateProviderExportOnDemand")]
+        public async Task<IActionResult> RunOnDemand([HttpTrigger(AuthorizationLevel.Anonymous, "post", Route = "bulk/generate")] HttpRequest req, ILogger log, CancellationToken ct)
+        {
+            var result = await GenerateProviderExport(log, ct);
 
-            log.LogInformation($"Started generation of {{{nameof(exportKey)}}}.", exportKey);
+            if (!result.Success)
+            {
+                return new ObjectResult(result.Message)
+                {
+                    StatusCode = StatusCodes.Status500InternalServerError
+                };
+            }
+
+            return new OkObjectResult(result.Message);
+        }
+
+        private async Task<GenerateProviderExportResult> GenerateProviderExport(ILogger log, CancellationToken ct)
+        {
+            var exportKey = ExportKey.FromUtcNow();
 
-            string export = null;
             try
             {
-                var generateStopwatch = Stopwatch.StartNew();
+                var blobClient = _blobStorageClient.GetBlobClient(exportKey);
+                var leaseClient = _blobStorageClient.GetBlobLeaseClient(blobClient);
+
+                Response<BlobLease> lease;
+                try
+                {
+                    lease = await leaseClient.AcquireAsync(TimeSpan.FromSeconds(60), cancellationToken: ct);
+                    log.LogInformation($"Acquired lease for {{{nameof(exportKey)}}}.", exportKey);
+                }
+                catch (RequestFailedException ex)
+                    when (ex.Status == 409 && ex.ErrorCode == BlobErrorCode.LeaseAlreadyPresent)
+                {
+                    log.LogWarning($"Failed to acquire lease for {{{nameof(exportKey)}}}.", exportKey);
+                    return GenerateProviderExportResult.FailedInProgress(exportKey);
+                }
+
+                var leaseRenewalTimer = new Timer(async _ =>
```

Comment:
Lease can only be acquired for between 15-60 seconds or infinite. 60 seconds isn't long enough, infinite is too risky, this is the cleanest [dirty] way I could think of.